File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.js
// banned_id별 후보 목록 만들기
// DFS로 하나씩 배정
// 결과는 Set으로 중복 제거

function solution(user_id, banned_id) {
    const result = new Set();

    // 후보를 "인덱스"로 저장
    const candidates = banned_id.map(ban => {
        const arr = [];
        for (let i = 0; i < user_id.length; i++) {
            if (isMatching(user_id[i], ban)) {
                arr.push(i);
            }
        }
        return arr;
    });

    const visited = Array(user_id.length).fill(false);

    function dfs(depth, selected) {
        if (depth === banned_id.length) {
            const sorted = [...selected].sort((a, b) => a - b).join(',');
            result.add(sorted);
            return;
        }

        for (let i = 0; i < candidates[depth].length; i++) {
            const userIdx = candidates[depth][i];

            if (visited[userIdx]) continue;

            visited[userIdx] = true;
            selected.push(userIdx);

            dfs(depth + 1, selected);

            selected.pop();
            visited[userIdx] = false;
        }
    }

    dfs(0, []);
    return result.size;
}

function isMatching(user, ban) {
    if (user.length !== ban.length) return false;

    for (let i = 0; i < ban.length; i++) {
        if (ban[i] === '*') continue;
        if (user[i] !== ban[i]) return false;
    }

    return true;
}

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.py
from itertools import product
from collections import defaultdict

def solution(user_id, banned_id):
    ban_ids = defaultdict(list)
    
    for i, ban in enumerate(banned_id):
        for user in user_id:
            if isBanned(ban, user):
                ban_ids[i].append(user)
    
    answer = set()
    
    # 각 banned_id 자리의 후보 리스트를 순서대로 꺼냄
    candidates = [ban_ids[i] for i in range(len(banned_id))]
    
    # 각 자리마다 후보 1명씩 뽑는 모든 경우
    for users in product(*candidates):
        # 같은 유저를 중복 선택하면 안 됨
        if len(set(users)) == len(banned_id):
            answer.add(frozenset(users))
    
    return len(answer)

def isBanned(a, b):
    if len(a) != len(b):
        return False
    
    for i in range(len(a)):
        if a[i] == '*':
            continue
        if a[i] != b[i]:
            return False
        
    return True